File: examples/node/list-an-item.mjs
#!/usr/bin/env node
/**
 * The whole path, in Node with no dependencies: photos in, a finished eBay
 * listing out, stopping before it goes live.
 *
 *   export VENDLISTS_KEY=vl_agent_your_key_here
 *   node examples/node/list-an-item.mjs ./photos/*.jpg
 *
 * It deliberately does NOT publish. Publishing puts a real item in front of
 * real buyers on the person's own eBay account, so it is one explicit call
 * further on, printed at the end for you to run when you mean it.
 */
import { readFile } from 'node:fs/promises';
import { basename, extname } from 'node:path';

const API = process.env.VENDLISTS_API ?? 'https://api.vendlists.com';
const KEY = process.env.VENDLISTS_KEY;

const CONTENT_TYPES = { '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg', '.png': 'image/png', '.webp': 'image/webp' };
const POLL_SECONDS = 6;
const GIVE_UP_AFTER_MINUTES = 10;

async function call(method, path, body) {
  const res = await fetch(`${API}${path}`, {
    method,
    headers: {
      Authorization: `Bearer ${KEY}`,
      ...(body ? { 'Content-Type': 'application/json' } : {}),
    },
    ...(body ? { body: JSON.stringify(body) } : {}),
  });
  const text = await res.text();
  const data = text ? JSON.parse(text) : null;
  if (res.status === 429) {
    // The API says how long to wait, so wait that long rather than guessing.
    const wait = Number(res.headers.get('retry-after') ?? 30);
    console.log(`  rate limited; waiting ${wait}s`);
    await sleep(wait * 1000);
    return call(method, path, body);
  }
  if (!res.ok) throw new Error(`${method} ${path} → ${res.status} ${text}`);
  return data;
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function main() {
  const files = process.argv.slice(2);
  if (!KEY) throw new Error('Set VENDLISTS_KEY to your vl_agent_… key.');
  if (files.length === 0) throw new Error('Pass one or more photo paths.');

  // 0. Where does this person stand? One call, before spending anything.
  const me = await call('GET', '/agent/me');
  console.log(`account: ${me.plan.name}, ${me.allowance.remaining} of ${me.allowance.included} listings left`);
  if (!me.ebay.connected) {
    console.log('eBay is not connected yet. The person connects it themselves at vendlists.com.');
  }
  if (me.allowance.nextListing.kind === 'blocked') {
    throw new Error(`No listing available: ${me.allowance.nextListing.reason}. See ${API}/agent/guide`);
  }

  // 1. The draft. Everything the person said about the item goes here: it is
  //    what the model cannot see in a photo.
  const { listingId } = await call('POST', '/listings', {
    additionalContext: process.env.ITEM_NOTES ?? 'Listed from the Vendlists agent kit example.',
    quantity: 1,
  });
  console.log(`draft: ${listingId}`);

  // 2. One upload URL per photo, then PUT the bytes with the same type.
  const urls = await call('POST', '/listings/upload-url', {
    listingId,
    files: files.map((file, index) => ({ contentType: contentTypeOf(file), index })),
  });
  await Promise.all(urls.map(async ({ uploadUrl }, i) => {
    const bytes = await readFile(files[i]);
    const put = await fetch(uploadUrl, {
      method: 'PUT',
      headers: { 'Content-Type': contentTypeOf(files[i]) },
      body: bytes,
    });
    if (!put.ok) throw new Error(`upload of ${basename(files[i])} → ${put.status}`);
    console.log(`  uploaded ${basename(files[i])}`);
  }));

  // 3. Vendlists writes it. This is the call that uses one of the month's listings.
  await call('POST', `/listings/${listingId}/generate`, {});
  console.log('writing the listing…');

  // 4. Poll only while it is being written, and never forever.
  const deadline = Date.now() + GIVE_UP_AFTER_MINUTES * 60_000;
  let listing;
  do {
    await sleep(POLL_SECONDS * 1000);
    listing = await call('GET', `/listings/${listingId}`);
    process.stdout.write(`  ${listing.status}\r`);
  } while (listing.status === 'processing' && Date.now() < deadline);

  if (listing.status === 'failed') {
    // `processingError` is a sentence written for the person, in their language.
    throw new Error(`${listing.processingErrorCode ?? 'FAILED'}: ${listing.processingError}`);
  }
  if (listing.status !== 'pending_review') {
    throw new Error(`Still ${listing.status} after ${GIVE_UP_AFTER_MINUTES} minutes.`);
  }

  // 5. Show the person what was written. They decide what happens next.
  console.log(`\n${listing.title}`);
  console.log(`  price:     ${(listing.price ?? listing.suggestedPrice ?? 0) / 100} ${listing.currency ?? ''}`);
  console.log(`  category:  ${listing.categoryName ?? '—'}`);
  console.log(`  condition: ${listing.condition ?? '—'}`);
  console.log(`  specifics: ${Object.keys(listing.itemSpecifics ?? {}).join(', ') || '—'}`);

  // 6. eBay's fee, which is eBay's and not Vendlists'. Show it before publishing.
  try {
    const fees = await call('POST', `/listings/${listingId}/channels/fees`, { action: 'quote' });
    console.log(`  eBay fee:  ${JSON.stringify(fees)}`);
  } catch (err) {
    console.log(`  eBay fee:  not quoted (${err.message.split('\n')[0]})`);
  }

  console.log(`
Review it at https://vendlists.com/dashboard/listings/${listingId}

When the person says yes, publish with:
  curl -X POST ${API}/ebay/publish/${listingId} \\
    -H "Authorization: Bearer $VENDLISTS_KEY" -H 'Content-Type: application/json' -d '{}'
`);
}

function contentTypeOf(file) {
  const type = CONTENT_TYPES[extname(file).toLowerCase()];
  if (!type) throw new Error(`${basename(file)}: photos must be jpeg, png or webp.`);
  return type;
}

main().catch((err) => {
  console.error(`\n${err.message}`);
  process.exit(1);
});

File: examples/python/list_an_item.py
#!/usr/bin/env python3
"""The whole path, in Python with the standard library only.

    export VENDLISTS_KEY=vl_agent_your_key_here
    python3 examples/python/list_an_item.py ./photos/*.jpg

It stops before publishing. Publishing puts a real item in front of real
buyers on the person's own eBay account, so it stays an explicit step.
"""
import json
import os
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path

API = os.environ.get("VENDLISTS_API", "https://api.vendlists.com")
KEY = os.environ.get("VENDLISTS_KEY")
CONTENT_TYPES = {".jpg": "image/jpeg", ".jpeg": "image/jpeg", ".png": "image/png", ".webp": "image/webp"}
POLL_SECONDS = 6
GIVE_UP_AFTER_MINUTES = 10


def call(method: str, path: str, body: dict | None = None):
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(f"{API}{path}", method=method, data=data)
    req.add_header("Authorization", f"Bearer {KEY}")
    if data:
        req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req, timeout=30) as res:
            raw = res.read().decode()
            return json.loads(raw) if raw else None
    except urllib.error.HTTPError as err:
        if err.code == 429:
            # The API says how long to wait; waiting less is how you stay throttled.
            wait = int(err.headers.get("Retry-After", "30"))
            print(f"  rate limited; waiting {wait}s")
            time.sleep(wait)
            return call(method, path, body)
        raise SystemExit(f"{method} {path} -> {err.code} {err.read().decode()[:400]}")


def content_type_of(file: str) -> str:
    suffix = Path(file).suffix.lower()
    if suffix not in CONTENT_TYPES:
        raise SystemExit(f"{Path(file).name}: photos must be jpeg, png or webp.")
    return CONTENT_TYPES[suffix]


def main(files: list[str]) -> None:
    if not KEY:
        raise SystemExit("Set VENDLISTS_KEY to your vl_agent_... key.")
    if not files:
        raise SystemExit("Pass one or more photo paths.")

    # 0. Where does this person stand, before anything is spent.
    me = call("GET", "/agent/me")
    allowance = me["allowance"]
    print(f"account: {me['plan']['name']}, {allowance['remaining']} of {allowance['included']} listings left")
    if not me["ebay"]["connected"]:
        print("eBay is not connected yet. The person connects it themselves at vendlists.com.")
    if allowance["nextListing"]["kind"] == "blocked":
        raise SystemExit(f"No listing available: {allowance['nextListing'].get('reason')}. See {API}/agent/guide")

    # 1. The draft, with everything the person said that a photo cannot show.
    listing_id = call("POST", "/listings", {
        "additionalContext": os.environ.get("ITEM_NOTES", "Listed from the Vendlists agent kit example."),
        "quantity": 1,
    })["listingId"]
    print(f"draft: {listing_id}")

    # 2. One upload URL per photo; PUT the bytes with the same content type.
    urls = call("POST", "/listings/upload-url", {
        "listingId": listing_id,
        "files": [{"contentType": content_type_of(f), "index": i} for i, f in enumerate(files)],
    })
    for file, entry in zip(files, urls):
        put = urllib.request.Request(entry["uploadUrl"], method="PUT", data=Path(file).read_bytes())
        put.add_header("Content-Type", content_type_of(file))
        with urllib.request.urlopen(put, timeout=60):
            print(f"  uploaded {Path(file).name}")

    # 3. The call that uses one of the month's listings.
    call("POST", f"/listings/{listing_id}/generate", {})
    print("writing the listing...")

    # 4. Poll only while it is being written, and never forever.
    deadline = time.time() + GIVE_UP_AFTER_MINUTES * 60
    listing = {"status": "processing"}
    while listing["status"] == "processing" and time.time() < deadline:
        time.sleep(POLL_SECONDS)
        listing = call("GET", f"/listings/{listing_id}")
        print(f"  {listing['status']}", end="\r")

    if listing["status"] == "failed":
        raise SystemExit(f"\n{listing.get('processingErrorCode', 'FAILED')}: {listing.get('processingError')}")
    if listing["status"] != "pending_review":
        raise SystemExit(f"\nStill {listing['status']} after {GIVE_UP_AFTER_MINUTES} minutes.")

    # 5. Show the person what was written; they decide what happens next.
    print(f"\n{listing.get('title')}")
    print(f"  price:     {(listing.get('price') or listing.get('suggestedPrice') or 0) / 100} {listing.get('currency', '')}")
    print(f"  category:  {listing.get('categoryName', '-')}")
    print(f"  condition: {listing.get('condition', '-')}")
    print(f"  specifics: {', '.join((listing.get('itemSpecifics') or {}).keys()) or '-'}")

    # 6. eBay's fee is eBay's, not Vendlists'. Show it before publishing.
    try:
        fees = call("POST", f"/listings/{listing_id}/channels/fees", {"action": "quote"})
        print(f"  eBay fee:  {json.dumps(fees)[:200]}")
    except SystemExit as err:
        print(f"  eBay fee:  not quoted ({err})")

    print(f"""
Review it at https://vendlists.com/dashboard/listings/{listing_id}

When the person says yes, publish with:
  curl -X POST {API}/ebay/publish/{listing_id} \\
    -H "Authorization: Bearer $VENDLISTS_KEY" -H 'Content-Type: application/json' -d '{{}}'
""")


if __name__ == "__main__":
    main(sys.argv[1:])
